pyxem/utils/_azimuthal_utils.py

```diff
@@ -57,13 +63,70 @@ def _slice_radial_integrate(
     """
     if mask is not None:
         img = img * np.logical_not(mask)
-    val = np.empty(slices.shape[0])
-    for i, (s, f) in enumerate(zip(slices, factors_slice)):
-        val[i] = np.sum(
-            img[s[0] : s[2], s[1] : s[3]]
-            * factors[f[0] : f[1]].reshape((s[2] - s[0], s[3] - s[1]))
+    val = np.empty((npt_rad, npt_azim))
+    for i in prange(len(factors_slice)):
+        ii, jj = i // npt_azim, i % npt_azim
+        val[ii, jj] = np.sum(
+            img[slices[i][0] : slices[i][2], slices[i][1] : slices[i][3]]
+            * factors[factors_slice[i][0] : factors_slice[i][1]].reshape(
+                (slices[i][2] - slices[i][0], slices[i][3] - slices[i][1])
+            )
         )
-    return val.reshape((npt_azim, npt_rad)).T
+    return val
+
+
+@cuda.jit
+def __slice_radial_integrate_cupy(
+    img, factors, factors_slice, slices, val
+):  # pragma: no cover
+    """Slice the image into small chunks and multiply by the factors.
+    Parameters
+    ----------
+    img: np.array
+        The image to be sliced
+    factors:
+        The factors to multiply the slices by
+    slices:
+        The slices to slice the image by
+    val:
+        The array to store the result in
+    Note
+    ----
+    This function is run by every single thread once!
+    """
+    tx = cuda.threadIdx.x  # current thread
+    bx = cuda.blockIdx.x  # Current block
+    bw = cuda.blockDim.x  # Should be equal to blocks!
+    x = tx + bx * bw
+    if x < val.shape[0]:  # account for slices out of range!
+        factors_ind = factors_slice[x]
+        current_slice = slices[x]
+        sum = 0
+        ind = 0
+        for i in range(current_slice[0], current_slice[2]):
+            for j in range(current_slice[1], current_slice[3]):
+                sum += factors[ind + factors_ind[0]] * img[i, j]
```

Comment:
Is there a way to know a-priori what data from `img` and `factors` are required for this calculation? Is there a way to group threads in such a way that adjacent threads use the same or overlapping information from `img` and `factors`? For example, by sorting your slices.

If you think not, then this is the best you will be able to do and likely the speed of the calculation will be limited simply by bandwidth from VRAM to compute cores.

If there is a way to group threads such that they use overlapping `factors` and/or `img` you could make use of shared memory arrays which has the potential to drastically speed up calculations as it saves you round trips from VRAM which are now occuring in each iteration of the double loop.

Just something to think about if you feel up to it.

Reply:
@din14970 that's a good point. 


The easy one is the factors. If you have some 3D block of diffraction patterns computing the same I(k,theta) and each thread operating along the navigation dimension that would be possible. That array is fairly small so I don't know if it will save that much time. It is getting accessed n times where n is the number of navigation positions so that might be worth it.

Each point in the img array is getting accessed on average about 2-3 times or so and shared memory would help there. Only problem is that the img array is a bit too large in most cases for shared memory?  Correct? You could split the image into chunks and possibly work that way but we are splitting hairs at that point.

I'll probably revisit this if someone actually wants to do this live. For now even a factor of 2 is probably not worth it especially when there is a lot more you could do with optimizing transfer to/from the gpu.